sys/kern/link_elf.c

```diff
@@ -1902,16 +1961,55 @@ elf_lookup(linker_file_t lf, Elf_Size symidx, int deps, Elf_Addr *res)
 		return (EINVAL);
 	}
 
-	if (elf_set_find(&set_pcpu_list, addr, &start, &base))
+	if (elf_set_find(&set_pcpu_list, addr, &start, &base)) {
 		addr = (vaddr_t)addr - (vaddr_t)start + base;
+#ifdef __CHERI_PURE_CAPABILITY__
+		addr = make_capability(sym, addr);
+#endif
+	}
 #ifdef VIMAGE
-	else if (elf_set_find(&set_vnet_list, addr, &start, &base))
+	else if (elf_set_find(&set_vnet_list, addr, &start, &base)) {
 		addr = (vaddr_t)addr - (vaddr_t)start + base;
+#ifdef __CHERI_PURE_CAPABILITY__
+		addr = make_capability(sym, addr);
+#endif
+	}
+#endif
+#ifdef __CHERI_PURE_CAPABILITY__
+	*res = (uintcap_t)addr;
+#else
+	*res = (vaddr_t)addr;
 #endif
```

Comment:
Currently this would work with `uintptr_t` everywhere, but if we ever want hybrid kernel modules with capability relocations then this structure makes more sense.

Reply:
Yeah, I thought about the uintptr_t route, but I was trying to be more consistent with the other #ifdef's above.

Also, I might rework some of this to make `ef->address` be a `ptraddr_t` as a followup change and add bounded caps used to derive caps via `cheri_setaddress` as another round of followup fixes.  In that case, this #ifdef would still be needed I think and perhaps more relevant.